### PR TITLE
Add daily active deposits metric

### PR DESCRIPTION
**Pull request by kalkov**
#### Summary

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex [hunk 1]
```diff
@@ -30,6 +30,26 @@ defmodule SanbaseWeb.Graphql.Resolvers.ClickhouseResolver do
       {:error, "Can't calculate network growth"}
   end
 
+  def daily_active_deposits(_root, args, _resolution) do
+    interval = DateTimeUtils.compound_duration_to_seconds(args.interval)
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: @tspenov @kalkov let's decide on a uniform way whether to transform the interval in the resolver or the function using it. My opinion is that the resolvers must be the cleanest function and should just call the modules.
One discrepancy that happens if we transform here is that all functions except these communicating with influxdb will transform the arguments, as the `GROUP BY` in influxb is the same string that is passed here.

Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: Also the `calibrate_interval` that is used a lot also returns a string `"<seconds>s"`

File: lib/sanbase/clickhouse/daily_active_deposits.ex [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  @type active_deposits :: %{
+          datetime: %DateTime{},
+          active_deposits: non_neg_integer()
+        }
+
+  @spec active_deposits(
+          String.t(),
+          %DateTime{},
+          %DateTime{},
+          Number.t()
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/daily_active_deposits.ex]: You can check in iex that `t Number.t` says `Could not load module Number, got: nofile`
Number type specs are on of:
- `non_neg_integer()`
- `pos_integer()`
- `neg_integer()`
- `integer()`
- `float()`

There's also `number()` which is an alias for `integer() | float()`

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex [hunk 3]
```diff
@@ -30,6 +30,26 @@ defmodule SanbaseWeb.Graphql.Resolvers.ClickhouseResolver do
...
+         {:ok, active_deposits} <-
+           DailyActiveDeposits.active_deposits(contract, args.from, args.to, interval) do
+      {:ok, active_deposits}
+    else
+      error ->
+        Logger.error("Can't calculate daily active deposits. Reason: #{inspect(error)}")
+
+        {:error, "Can't calculate daily active deposits"}
+    end
+  rescue
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: I really dislike this pattern we have at many places. I suggest to leave it like this but create a favro card that handles this on a middleware level (that we could probably plug for all queries?).
These 2 blogposts are on that topic:
1. Introducing the problem and solving it this way: https://tech.decisiv.com/handling-elixir-exceptions-in-absinthe-9656a9108c2d
2. Solving it with a middleware: https://tech.decisiv.com/handling-elixir-exceptions-in-absinthe-part-2-using-middleware-d20e7df7cc62

The proposed solution seems to wrap all middlewares in such safety box, though I think we'll need it only in some places and let other queries just crash.

The main place where we must handle all errors is the `all_*` queries, because if a single field there fails, the whole request fails and the overview page does not load.

File: lib/sanbase_web/graphql/schema.ex [hunk 4]
```diff
@@ -416,6 +416,21 @@ defmodule SanbaseWeb.Graphql.Schema do
...
 
+    @desc ~s"""
+    Fetch daily active deposits for a project within a given time period.
+    Projects are referred to by a unique identifier (slug).
+    """
+    field :daily_active_deposits, list_of(:active_deposits) do
+      arg(:slug, non_null(:string))
+      arg(:from, non_null(:datetime))
+      arg(:to, non_null(:datetime))
+      arg(:interval, :string, default_value: "1d")
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/schema.ex]: `interval` can also have a default value of `""`. In that case you can use the `calibrate_interval` function that is crafted so it transforms the interval in such a way that in the time period `from - to` there are a specific number of data points, so it looks good on the chart that the frontend draws.

Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/schema.ex]: Manually calling and handling `calibrate_interval` in each resolver is also not nice, as it's very verbose. This is again one thing we could probably modify with a middleware.
Saying this I figured out that we have an issue - at some places we handle `""` and set a specific value to it, but in all calculations for complexity the `""` interval is handled as if it is `"1d"`

File: test/sanbase_web/graphql/clickhouse/active_deposits_test.exs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+  setup do
+    project = insert(:project, %{main_contract_address: "0x123"})
+
+    [
+      project: project
+    ]
+  end
+
+  test "active deposits works", context do
```
Comment (IvanIvanoff) [inline test/sanbase_web/graphql/clickhouse/active_deposits_test.exs]: you can locally run `MIX_TEST=test mix coveralls.html` and check the `cover/<name>.html` file so you can see what other parts you need to cover in your test (the error cases and empty cases, for example).

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex (hunk 1 above)
Comment (kalkov) [reply lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: @IvanIvanoff Good arguments, my vote goes to not transform in the resolver.

File: lib/sanbase/clickhouse/daily_active_deposits.ex (hunk 2 above)
Comment (kalkov) [reply lib/sanbase/clickhouse/daily_active_deposits.ex]: Tried to guess it, failed.

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex (hunk 1 above)
Comment (tspenov) [reply lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: Also agree on keeping short and sweat the resolvers

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex [hunk 6]
```diff
@@ -6,7 +6,7 @@ defmodule SanbaseWeb.Graphql.Resolvers.ClickhouseResolver do
   alias Sanbase.Clickhouse.Erc20Transfers
   alias Sanbase.Model.Project
 
-  alias Sanbase.Clickhouse.NetworkGrowth
+  alias Sanbase.Clickhouse.{ NetworkGrowth, DailyActiveDeposits }
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: `{NetworkGrowth, DailyActiveDeposits}` extra space on both sides.
We don't have a style guide  but we run `mix format` before commit - it should fix these style things

Comment (tspenov) [reply lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: here is a snippet (that I've borrowed form @IvanIvanoff ) to put in `.git/hooks/pre-commit`
```bash
#!/bin/bash

cd `git rev-parse --show-toplevel`
echo "Running pre-commit hook: Check elixir formatting with mix format"
mix format --check-formatted
if [ $? == 1 ]; then
  echo "mix format --check-formatted found formatting issues."
  exit 1
fi
```

File: test/sanbase_web/graphql/clickhouse/active_deposits_test.exs (hunk 5 above)
Comment (valo) [reply test/sanbase_web/graphql/clickhouse/active_deposits_test.exs]: I am wondering if we should plug the coverage stats to GitHub

File: lib/sanbase_web/graphql/schema.ex (hunk 4 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/schema.ex]: On the other hand, I think it's ok to leave the complexity calculations like this, as the current behavior is more restrictive and will not allow some queries, which is ok.

File: test/sanbase_web/graphql/clickhouse/active_deposits_test.exs (hunk 5 above)
Comment (kalkov) [reply test/sanbase_web/graphql/clickhouse/active_deposits_test.exs]: @valo You mean to add a coverage badge?

File: lib/sanbase/clickhouse/daily_active_deposits.ex [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  alias Sanbase.DateTimeUtils
+
+  @type active_deposits :: %{
+          datetime: %DateTime{},
+          active_deposits: non_neg_integer()
+        }
+
+  @spec active_deposits(
+          String.t(),
+          %DateTime{},
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/daily_active_deposits.ex]: for some reasons we have this `%DateTime{}` in the specs, though there is the `DateTime.t()` available

File: test/sanbase_web/graphql/clickhouse/active_deposits_test.exs [hunk 8]
```diff
@@ -0,0 +1,55 @@
+defmodule SanbaseWeb.Graphql.ClickhouseActiveDepositsTest do
+  use SanbaseWeb.ConnCase
+  require Sanbase.ClickhouseRepo
```
Comment (IvanIvanoff) [inline test/sanbase_web/graphql/clickhouse/active_deposits_test.exs]: Here you're only mocking it so there should be no need to `require` it. `require` is needed when you use a macro from that module

File: test/sanbase_web/graphql/clickhouse/active_deposits_test.exs (hunk 5 above)
Comment (valo) [reply test/sanbase_web/graphql/clickhouse/active_deposits_test.exs]: Sure, let's try it. This service https://coveralls.io supports open source repos for free, so we can try it too. I think that will be a separate task and not included in this PR.

File: lib/sanbase/clickhouse/daily_active_deposits.ex [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+          %DateTime{},
+          String.t()
+        ) :: {:ok, list(active_deposits)} | {:error, String.t()}
+  def active_deposits(contract, from, to, interval) do
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/daily_active_deposits.ex]: You can add `active_deposits_query` function

File: test/sanbase_web/graphql/clickhouse/daily_active_deposits_test.exs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    with_mock DailyActiveDeposits,
+      active_deposits: fn _, _, _, _ ->
+        {:ok,
+         [
+           %{active_deposits: 100, datetime: from_iso8601!("2019-01-01T00:00:00Z")},
+           %{active_deposits: 200, datetime: from_iso8601!("2019-01-02T00:00:00Z")}
+         ]}
+      end do
+      response = execute_query(context)
+      ratios = parse_response(response)
```
Comment (tspenov) [inline test/sanbase_web/graphql/clickhouse/daily_active_deposits_test.exs]: this should be called probably `deposits`

**Comment by tspenov** [comment]
👍 LGTM - small comment in test

File: lib/sanbase/clickhouse/daily_active_deposits.ex [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  Uses ClickHouse to calculate daily active deposits
+  """
+
+  require Sanbase.ClickhouseRepo, as: ClickhouseRepo
+  import Sanbase.Math, only: [to_integer: 1]
+
+  alias Sanbase.DateTimeUtils
+
+  @type active_deposits :: %{
+          datetime: %DateTime{},
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/daily_active_deposits.ex]: Here the datetime type can be changed, too

File: lib/sanbase/clickhouse/daily_active_deposits.ex [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        SUM(total_addresses) AS value
+      FROM daily_active_deposits
+      PREWHERE
+        contract = ?3 AND
+        dt >= toDateTime(?4) AND
+        dt <= toDateTime(?5)
+      GROUP BY time
+    )
+    GROUP BY dt
+    ORDER BY dt
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/daily_active_deposits.ex]: I am a little bit confused by the redefinitions of `dt` and `time`. Is it possible to do something like `toUnixTimestamp(time) as time`?

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex [hunk 13]
```diff
@@ -36,6 +39,34 @@ defmodule SanbaseWeb.Graphql.Resolvers.ClickhouseResolver do
...
+    else
+      {:error, error} ->
+        Logger.warn(
+          "Can't calculate daily active deposits for project with coinmarketcap_id: #{args.slug}. Reason: #{
+            inspect(error)
+          }"
+        )
+
+        {:error,
+         "Can't calculate daily active deposits for project with coinmarketcap_id: #{args.slug}."}
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]:
```elixir
error_msg = "Can't calculate daily active deposits for project with coinmarketcap_id: #{args.slug}."
Logger.warn(error_msg <> " Reason: #{inspect(error)}"
{:error, error_msg}
```

File: lib/sanbase_web/graphql/schema.ex [hunk 14]
```diff
@@ -428,6 +428,21 @@ defmodule SanbaseWeb.Graphql.Schema do
...
 
+    @desc ~s"""
+    Fetch daily active deposits for a project within a given time period.
+    Projects are referred to by a unique identifier (slug).
+    """
+    field :daily_active_deposits, list_of(:active_deposits) do
+      arg(:slug, non_null(:string))
+      arg(:from, non_null(:datetime))
+      arg(:to, non_null(:datetime))
+      arg(:interval, non_null(:string), default_value: "1d")
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/schema.ex]: GraphiQL is confused by non_null + default values. Not sure if it's an issue with graphiql though

File: lib/sanbase_web/graphql/schema.ex [hunk 15]
```diff
@@ -428,6 +428,21 @@ defmodule SanbaseWeb.Graphql.Schema do
...
+
+      complexity(&Complexity.from_to_interval/3)
+      middleware(ApiTimeframeRestriction, %{allow_historical_data: true})
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/schema.ex]: Is this the desired access or we want to cut off the historical data, too?

File: lib/sanbase_web/graphql/schema/clickhouse_types.ex [hunk 16]
```diff
@@ -1,18 +1,23 @@
 defmodule SanbaseWeb.Graphql.ClickhouseTypes do
   use Absinthe.Schema.Notation
 
-  object :historical_balance do
+  object :active_deposits do
     field(:datetime, non_null(:datetime))
-    field(:balance, :float)
+    field(:active_deposits, :integer)
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/schema/clickhouse_types.ex]: If we're filling empty values with `0` (the correct behavior) then it can be non_null here

File: test/sanbase/clickhouse/daily_active_deposits_test.exs [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  import Sanbase.DateTimeUtils, only: [from_iso8601_to_unix!: 1, from_iso8601!: 1]
+  alias Sanbase.Clickhouse.DailyActiveDeposits
+  require Sanbase.ClickhouseRepo
+
+  setup do
+    project =
+      insert(:project, %{
+        coinmarketcap_id: "ethereum",
+        ticker: "ETH",
+        main_contract_address: "0x123"
```
Comment (IvanIvanoff) [inline test/sanbase/clickhouse/daily_active_deposits_test.exs]: That is confusing because when you call `contract_info` on `"ethereum"` it has a different effect - it does not check the contract field but returns `ETH` directly.

File: lib/sanbase/clickhouse/daily_active_deposits.ex [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+      }
+    end)
+  end
+
+  defp active_deposits_query(contract, from, to, interval) do
+    from_datetime_unix = DateTime.to_unix(from)
+    to_datetime_unix = DateTime.to_unix(to)
+    interval = DateTimeUtils.compound_duration_to_seconds(interval)
+    span = div(to_datetime_unix - from_datetime_unix, interval) |> max(1)
+
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/daily_active_deposits.ex]: You'll need to downcase the contract as this is the behavior in ClickHouse. We have (mostly) downcased addresses in our DB but for example the `contract_info` for ethereum would give `ETH` while the clickhouse contract for it is `eth`

File: lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex [hunk 19]
```diff
@@ -36,6 +39,34 @@ defmodule SanbaseWeb.Graphql.Resolvers.ClickhouseResolver do
     end
   end
 
+  def daily_active_deposits(_root, args, _resolution) do
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/clickhouse_resolver.ex]: In most places we pattern match the args in the function head. It is arguably nicer to read and you'll fail early with a more meaningful error when not correct params are given.

File: lib/sanbase_web/graphql/schema.ex (hunk 15 above)
Comment (kalkov) [reply lib/sanbase_web/graphql/schema.ex]: I think not, fixed

File: lib/sanbase/clickhouse/daily_active_deposits.ex (hunk 12 above)
Comment (kalkov) [reply lib/sanbase/clickhouse/daily_active_deposits.ex]: Addressed, please check the fix
